util/patch/patch_test.go

```diff
@@ -997,3 +999,1683 @@ func TestNewHelperNil(t *testing.T) {
+			g.Expect(env.Status().Update(ctx, objCopy)).To(Succeed())
+
+			t.Log("Validating that the local object's resource version is behind")
+			g.Expect(obj.ResourceVersion).NotTo(Equal(objCopy.ResourceVersion))
+
+			t.Log("Creating a new patch helper")
+			patcher, err := NewHelper(obj, env)
+			g.Expect(err).ToNot(HaveOccurred())
+
+			t.Log("Marking Ready=True")
+			conditions.MarkTrue(obj, clusterv1.ReadyCondition)
+
+			t.Log("Patching the object")
+			g.Expect(patcher.Patch(ctx, obj, WithForceOverwriteConditions{})).To(Succeed())
+
+			t.Log("Validating the object has been updated")
+			readyBefore := conditions.Get(obj, clusterv1.ReadyCondition)
+			g.Eventually(func() clusterv1.Condition {
+				objAfter := obj.DeepCopy()
+				if err := env.Get(ctx, key, objAfter); err != nil {
+					return clusterv1.Condition{}
+				}
+
+				return *conditions.Get(objAfter, clusterv1.ReadyCondition)
+			}, timeout).Should(conditions.MatchCondition(*readyBefore))
+		})
+	})
+
+	t.Run("Should patch conditions on a v1beta1 object with both clusterv1.conditions and metav1.conditions (phase 1)", func(t *testing.T) {
+		obj := &builder.Phase1Obj{
+			ObjectMeta: metav1.ObjectMeta{
+				GenerateName: "test-",
+				Namespace:    metav1.NamespaceDefault,
+			},
+		}
+
+		t.Run("should mark it ready", func(t *testing.T) {
+			g := NewWithT(t)
+
+			obj := obj.DeepCopy()
+
+			t.Log("Creating the object")
+			err := env.Create(ctx, obj)
+			g.Expect(err).To(Succeed())
+			defer func() {
+				g.Expect(env.Delete(ctx, obj)).To(Succeed())
+			}()
+			key := client.ObjectKey{Name: obj.Name, Namespace: obj.Namespace}
+
+			t.Log("Checking that the object has been created")
+			g.Eventually(func() error {
+				obj := obj.DeepCopy()
+				return env.Get(ctx, key, obj)
+			}).Should(Succeed())
+
+			t.Log("Creating a new patch helper")
+			patcher, err := NewHelper(obj, env)
+			g.Expect(err).ToNot(HaveOccurred())
+
+			t.Log("Marking clusterv1.conditions and metav1.conditions Ready=True")
+			conditions.MarkTrue(obj, clusterv1.ReadyCondition)
+			err = v1beta2conditions.Set(obj, metav1.Condition{Type: "Ready", Status: metav1.ConditionTrue, Reason: "AllGood"})
+			g.Expect(err).ToNot(HaveOccurred())
+
+			t.Log("Patching the object")
+			g.Expect(patcher.Patch(ctx, obj)).To(Succeed())
+
+			t.Log("Validating the object has been updated")
+			g.Eventually(func() clusterv1.Conditions {
+				objAfter := obj.DeepCopy()
+				if err := env.Get(ctx, key, objAfter); err != nil {
+					return clusterv1.Conditions{}
+				}
+				return objAfter.Status.Conditions
+			}, timeout).Should(conditions.MatchConditions(obj.Status.Conditions))
+			g.Eventually(func() []metav1.Condition {
+				objAfter := obj.DeepCopy()
+				if err := env.Get(ctx, key, objAfter); err != nil {
+					return nil
+				}
+				return objAfter.Status.V1Beta2.Conditions
+			}, timeout).Should(v1beta2conditions.MatchConditions(obj.Status.V1Beta2.Conditions, v1beta2conditions.IgnoreLastTransitionTime(true)))
```

Comment:
I would have expected the last transition time to be equal.

Especially because patcher.Patch should modify obj. So it's also not clear to me if we are testing the right thing (obj is not actually the obj before we call patch, it's already the one from afterwards) Maybe we should use deepcopy to create a beforeObj before we call Patch and compare with that
(Probably same in many more cases here)